src/import/pull-common.c

```diff
@@ -637,6 +640,99 @@ int pull_job_restart_with_sha256sum(PullJob *j, char **ret) {
         return 1;
 }
 
+int signature_style_from_url(const char *url, SignatureStyle *ret) {
+        _cleanup_free_ char *last = NULL;
+        int r;
+
+        assert(url);
+        assert(ret);
+
+        /* Determines which kind of signature style is appropriate for this url */
+
+        r = import_url_last_component(url, &last);
+        if (r < 0)
+                return r;
+
+        if (streq(last, "SHA256SUMS.gpg")) {
+                *ret = SIGNATURE_GPG_PER_DIRECTORY;
+                return 0;
+        }
+
+        if (streq(last, "SHA256SUMS.asc")) {
+                *ret = SIGNATURE_ASC_PER_DIRECTORY;
+                return 0;
+        }
+
+        if (endswith(last, ".sha256.gpg")) {
+                *ret = SIGNATURE_GPG_PER_FILE;
+                return 0;
+        }
+
+        if (endswith(last, ".sha256.asc")) {
+                *ret = SIGNATURE_ASC_PER_FILE;
+                return 0;
+        }
+
+        return -EINVAL;
+}
+
+int pull_job_restart_with_signature(PullJob *j, char **ret) {
+        SignatureStyle style;
+        int r;
+
+        assert(j);
+
+        /* Generic implementation of a PullJobNotFound handler, that restarts the job requesting a different
+         * signature file. After the initial file, additional *.sha256.gpg, SHA256SUMS.gpg and SHA256SUMS.asc
+         * are tried in this order. */
+
+        r = signature_style_from_url(j->url, &style);
+        if (r < 0)
+                return log_error_errno(r, "Failed to determine signature style of URL '%s': %m", j->url);
+
+        switch(style) {
+
+        case SIGNATURE_ASC_PER_DIRECTORY: /* Nothing to do anymore */
+                return 0;
+
+        case SIGNATURE_ASC_PER_FILE: { /* Try .sha256.gpg next */
+                _cleanup_free_ char *last = NULL, *tmp = NULL;
+
+                log_debug("Got 404 for %s, now trying to get .sha256.gpg instead.", j->url);
+
+                r = import_url_last_component(j->url, &last);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get last URL part: %m");
+
+                tmp = strreplace(last, ".asc", ".gpg");
```

Comment:
OOM check is missing.

more importanlty, if i call my image file "upwards.ascending.stairs.asc.gpg" I'll end up with some weird url here...

Reply:
maybe change signature_style_from_url() to return the correct "basename" of the specified path too. After all it does the endswith(), and that already returns where the suffix starts...